From 21de5a1aa98171b257c8cbbce154d295ed69d94c Mon Sep 17 00:00:00 2001
From: Jes Sorensen <Jes.Sorensen@redhat.com>
Date: Mon, 1 Sep 2014 09:25:38 -0400
Subject: [md] test mddev->flags more safely in md_check_recovery

Message-id: <1409563584-9000-37-git-send-email-Jes.Sorensen@redhat.com>
Patchwork-id: 89985
O-Subject: [RHEL7.1 PATCH 36/82] md: test mddev->flags more safely in md_check_recovery.
Bugzilla: 1085530
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

From: NeilBrown <neilb@suse.de>

commit 7a0a5355cbc71efa md: Don't test all of mddev->flags at once.
made most tests on mddev->flags safer, but missed one.

When
commit 260fa034ef7a4ff8b7306 md: avoid deadlock when dirty buffers during md_stop.
added MD_STILL_CLOSED, this caused md_check_recovery to misbehave.
It can think there is something to do but find nothing.  This can
lead to the md thread spinning during array shutdown.

https://bugzilla.kernel.org/show_bug.cgi?id=65721

Reported-and-tested-by: Richard W.M. Jones <rjones@redhat.com>
Fixes: 260fa034ef7a4ff8b7306
Cc: stable@vger.kernel.org (3.12)
Signed-off-by: NeilBrown <neilb@suse.de>
(cherry picked from commit 142d44c310819e1965ca70b4d55d7679f5797e25)
Signed-off-by: Jes Sorensen <Jes.Sorensen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/md.c b/drivers/md/md.c
index e396c27..b5f1748 100644
--- a/drivers/md/md.c
+++ b/drivers/md/md.c
@@ -7773,7 +7773,7 @@ void md_check_recovery(struct mddev *mddev)
  if (mddev->ro && !test_bit(MD_RECOVERY_NEEDED, &mddev->recovery))
   return;
  if ( ! (
-  (mddev->flags & ~ (1<<MD_CHANGE_PENDING)) ||
+  (mddev->flags & MD_UPDATE_SB_FLAGS & ~ (1<<MD_CHANGE_PENDING)) ||
   test_bit(MD_RECOVERY_NEEDED, &mddev->recovery) ||
   test_bit(MD_RECOVERY_DONE, &mddev->recovery) ||
   (mddev->external == 0 && mddev->safemode == 1) ||
-- 
1.7.1